ChannelClosed exception when daemon is running for a long time

Potentially related to #1335 

I was running the daemon with twelve workers on theossrv5 and after 12 hours of heavy workchain submissions, a workchain crashed with the following stack trace:

```
  File "/home/aiida/.virtualenvs/aiida_sdb/src/plumpy/plumpy/base_process.py", line 569, in _run_task
    return fn(*args, **kwargs)
  File "/home/aiida/.virtualenvs/aiida_sdb/src/plumpy/plumpy/base_process.py", line 304, in _run
    self._action_command(command)
  File "/home/aiida/.virtualenvs/aiida_sdb/src/plumpy/plumpy/base_process.py", line 316, in _action_command
    self.transition_to(ProcessState.RUNNING, command.continue_fn, *command.args)
  File "/home/aiida/.virtualenvs/aiida_sdb/src/plumpy/plumpy/base/state_machine.py", line 148, in transition_to
    self.state_machine.transition_to(state, *args, **kwargs)
  File "/home/aiida/.virtualenvs/aiida_sdb/src/plumpy/plumpy/processes.py", line 446, in transition_to
    self.on_entered(initial_state)
  File "/home/aiida/.virtualenvs/aiida_sdb/src/plumpy/plumpy/processes.py", line 393, in on_entered
    subject="state_changed.{}.{}".format(from_label, self.state.value)
  File "/home/aiida/.virtualenvs/aiida_sdb/src/kiwipy/kiwipy/rmq/communicator.py", line 329, in broadcast_send
    return self._message_publisher.broadcast_send(body, sender, subject, correlation_id)
  File "/home/aiida/.virtualenvs/aiida_sdb/src/kiwipy/kiwipy/rmq/communicator.py", line 50, in broadcast_send
    self.action_message(message)
  File "/home/aiida/.virtualenvs/aiida_sdb/src/kiwipy/kiwipy/rmq/messages.py", line 303, in action_message
    message.send(self)
  File "/home/aiida/.virtualenvs/aiida_sdb/src/kiwipy/kiwipy/rmq/messages.py", line 111, in send
    publisher.publish_msg(self.message, defaults.BROADCAST_TOPIC)
  File "/home/aiida/.virtualenvs/aiida_sdb/src/kiwipy/kiwipy/rmq/messages.py", line 326, in publish_msg
    mandatory=mandatory
  File "/home/aiida/.virtualenvs/aiida_sdb/src/kiwipy/kiwipy/rmq/messages.py", line 342, in do_publish
    publish_channel.publish(*args, **kwargs)
  File "/home/aiida/.virtualenvs/aiida_sdb/lib/python2.7/site-packages/pika/adapters/blocking_connection.py", line 2127, in publish
    self._flush_output(self._message_confirmation_result.is_ready)
  File "/home/aiida/.virtualenvs/aiida_sdb/lib/python2.7/site-packages/pika/adapters/blocking_connection.py", line 1250, in _flush_output
    *waiters)
  File "/home/aiida/.virtualenvs/aiida_sdb/lib/python2.7/site-packages/pika/adapters/blocking_connection.py", line 455, in _flush_output
    self._impl.ioloop.poll()
  File "/home/aiida/.virtualenvs/aiida_sdb/lib/python2.7/site-packages/pika/adapters/select_connection.py", line 245, in poll
    self._poller.poll()
  File "/home/aiida/.virtualenvs/aiida_sdb/lib/python2.7/site-packages/pika/adapters/select_connection.py", line 935, in poll
    self._dispatch_fd_events(fd_event_map)
  File "/home/aiida/.virtualenvs/aiida_sdb/lib/python2.7/site-packages/pika/adapters/select_connection.py", line 625, in _dispatch_fd_events
    handler(fileno, events)
  File "/home/aiida/.virtualenvs/aiida_sdb/lib/python2.7/site-packages/pika/adapters/base_connection.py", line 395, in _handle_events
    self._handle_read()
  File "/home/aiida/.virtualenvs/aiida_sdb/lib/python2.7/site-packages/pika/adapters/base_connection.py", line 449, in _handle_read
    self._on_data_available(data)
  File "/home/aiida/.virtualenvs/aiida_sdb/lib/python2.7/site-packages/pika/connection.py", line 1938, in _on_data_available
    self._process_frame(frame_value)
  File "/home/aiida/.virtualenvs/aiida_sdb/lib/python2.7/site-packages/pika/connection.py", line 2059, in _process_frame
    if self._process_callbacks(frame_value):
  File "/home/aiida/.virtualenvs/aiida_sdb/lib/python2.7/site-packages/pika/connection.py", line 2040, in _process_callbacks
    frame_value)  # Args
  File "/home/aiida/.virtualenvs/aiida_sdb/lib/python2.7/site-packages/pika/callback.py", line 60, in wrapper
    return function(*tuple(args), **kwargs)
  File "/home/aiida/.virtualenvs/aiida_sdb/lib/python2.7/site-packages/pika/callback.py", line 92, in wrapper
    return function(*args, **kwargs)
  File "/home/aiida/.virtualenvs/aiida_sdb/lib/python2.7/site-packages/pika/callback.py", line 236, in process
    callback(*args, **keywords)
  File "/home/aiida/.virtualenvs/aiida_sdb/lib/python2.7/site-packages/pika/adapters/blocking_connection.py", line 1316, in _on_channel_closed
    method.reply_text)
ChannelClosed: (404, "NOT_FOUND - no exchange 'aiida-69b2078be6d74abd97446c7371983e62.messages' in vhost '/'")
```

I don't think time is the relevant factor here -- I managed to trigger the same exception almost instantaneously, by running a very heavy calculation (uses a few GB of input data). Traceback:

```
04/03/2018 02:38:40 PM <5541> pika.adapters.base_connection: [ERROR] Error event 25, None
04/03/2018 02:38:40 PM <5541> pika.adapters.base_connection: [CRITICAL] Tried to handle an error where no error existed
04/03/2018 02:38:40 PM <5541> pika.callback: [ERROR] Calling <bound method BlockingChannel._on_channel_closed of <BlockingChannel impl=<Channel number=1 CLOSED conn=<SelectConnection OPEN socket=('127.0.0.1', 35110)->('127.0.0.1', 5672) params=<URLParameters host=127.0.0.1 port=5672 virtual_host=/ ssl=False>>>>> for "1:Channel.Close" failed
Traceback (most recent call last):
  File "/home/greschd/.virtualenvs/cd3as2_aiida/local/lib/python2.7/site-packages/pika/callback.py", line 236, in process
    callback(*args, **keywords)
  File "/home/greschd/.virtualenvs/cd3as2_aiida/local/lib/python2.7/site-packages/pika/adapters/blocking_connection.py", line 1316, in _on_channel_closed
    method.reply_text)
ChannelClosed: (404, "NOT_FOUND - no exchange 'aiida-d26c044f49e14a1ea8e3f4d23d1170cb.messages' in vhost '/'")
04/03/2018 02:38:40 PM <5541> aiida.aiida_wannier90.calculations.Wannier90Calculation: [REPORT] [788|JobProcess_aiida_wannier90.calculations.Wannier90Calculation|on_except]: Traceback (most recent call last):
  File "/media/greschd/large_data/projects/Cd3As2/tbextraction/source/aiida_core/aiida/work/job_processes.py", line 242, in action_command
    self.scheduler_update()
  File "/media/greschd/large_data/projects/Cd3As2/tbextraction/source/aiida_core/aiida/work/job_processes.py", line 289, in scheduler_update
    data=UPDATE_SCHEDULER_COMMAND)
  File "/home/greschd/.virtualenvs/cd3as2_aiida/src/plumpy/plumpy/base/state_machine.py", line 148, in transition_to
    self.state_machine.transition_to(state, *args, **kwargs)
  File "/home/greschd/.virtualenvs/cd3as2_aiida/src/plumpy/plumpy/processes.py", line 458, in transition_to
    self.on_entered(initial_state)
  File "/home/greschd/.virtualenvs/cd3as2_aiida/src/plumpy/plumpy/processes.py", line 402, in on_entered
    subject='state_changed.{}.{}'.format(from_label, self.state.value)
  File "/home/greschd/.virtualenvs/cd3as2_aiida/src/kiwipy/kiwipy/rmq/communicator.py", line 329, in broadcast_send
    return self._message_publisher.broadcast_send(body, sender, subject, correlation_id)
  File "/home/greschd/.virtualenvs/cd3as2_aiida/src/kiwipy/kiwipy/rmq/communicator.py", line 50, in broadcast_send
    self.action_message(message)
  File "/home/greschd/.virtualenvs/cd3as2_aiida/src/kiwipy/kiwipy/rmq/messages.py", line 303, in action_message
    message.send(self)
  File "/home/greschd/.virtualenvs/cd3as2_aiida/src/kiwipy/kiwipy/rmq/messages.py", line 111, in send
    publisher.publish_msg(self.message, defaults.BROADCAST_TOPIC)
  File "/home/greschd/.virtualenvs/cd3as2_aiida/src/kiwipy/kiwipy/rmq/messages.py", line 326, in publish_msg
    mandatory=mandatory
  File "/home/greschd/.virtualenvs/cd3as2_aiida/src/kiwipy/kiwipy/rmq/messages.py", line 342, in do_publish
    publish_channel.publish(*args, **kwargs)
  File "/home/greschd/.virtualenvs/cd3as2_aiida/local/lib/python2.7/site-packages/pika/adapters/blocking_connection.py", line 2127, in publish
    self._flush_output(self._message_confirmation_result.is_ready)
  File "/home/greschd/.virtualenvs/cd3as2_aiida/local/lib/python2.7/site-packages/pika/adapters/blocking_connection.py", line 1250, in _flush_output
    *waiters)
  File "/home/greschd/.virtualenvs/cd3as2_aiida/local/lib/python2.7/site-packages/pika/adapters/blocking_connection.py", line 455, in _flush_output
    self._impl.ioloop.poll()
  File "/home/greschd/.virtualenvs/cd3as2_aiida/local/lib/python2.7/site-packages/pika/adapters/select_connection.py", line 245, in poll
    self._poller.poll()
  File "/home/greschd/.virtualenvs/cd3as2_aiida/local/lib/python2.7/site-packages/pika/adapters/select_connection.py", line 935, in poll
    self._dispatch_fd_events(fd_event_map)
  File "/home/greschd/.virtualenvs/cd3as2_aiida/local/lib/python2.7/site-packages/pika/adapters/select_connection.py", line 625, in _dispatch_fd_events
    handler(fileno, events)
  File "/home/greschd/.virtualenvs/cd3as2_aiida/local/lib/python2.7/site-packages/pika/adapters/base_connection.py", line 395, in _handle_events
    self._handle_read()
  File "/home/greschd/.virtualenvs/cd3as2_aiida/local/lib/python2.7/site-packages/pika/adapters/base_connection.py", line 449, in _handle_read
    self._on_data_available(data)
  File "/home/greschd/.virtualenvs/cd3as2_aiida/local/lib/python2.7/site-packages/pika/connection.py", line 1938, in _on_data_available
    self._process_frame(frame_value)
  File "/home/greschd/.virtualenvs/cd3as2_aiida/local/lib/python2.7/site-packages/pika/connection.py", line 2059, in _process_frame
    if self._process_callbacks(frame_value):
  File "/home/greschd/.virtualenvs/cd3as2_aiida/local/lib/python2.7/site-packages/pika/connection.py", line 2040, in _process_callbacks
    frame_value)  # Args
  File "/home/greschd/.virtualenvs/cd3as2_aiida/local/lib/python2.7/site-packages/pika/callback.py", line 60, in wrapper
    return function(*tuple(args), **kwargs)
  File "/home/greschd/.virtualenvs/cd3as2_aiida/local/lib/python2.7/site-packages/pika/callback.py", line 92, in wrapper
    return function(*args, **kwargs)
  File "/home/greschd/.virtualenvs/cd3as2_aiida/local/lib/python2.7/site-packages/pika/callback.py", line 236, in process
    callback(*args, **keywords)
  File "/home/greschd/.virtualenvs/cd3as2_aiida/local/lib/python2.7/site-packages/pika/adapters/blocking_connection.py", line 1316, in _on_channel_closed
    method.reply_text)
ChannelClosed: (404, "NOT_FOUND - no exchange 'aiida-d26c044f49e14a1ea8e3f4d23d1170cb.messages' in vhost '/'")
```

Somewhere down the line I also get 

```
04/03/2018 02:38:42 PM <5541> tornado.application: [ERROR] Future exception was never retrieved: Traceback (most recent call last):
  File "/home/greschd/.virtualenvs/cd3as2_aiida/local/lib/python2.7/site-packages/tornado/gen.py", line 1063, in run
    yielded = self.gen.throw(*exc_info)
  File "/home/greschd/.virtualenvs/cd3as2_aiida/src/kiwipy/kiwipy/rmq/tasks.py", line 140, in _on_task
    self._task_finished(method.delivery_tag, props.correlation_id, props.reply_to, response)
  File "/home/greschd/.virtualenvs/cd3as2_aiida/src/kiwipy/kiwipy/rmq/tasks.py", line 157, in _task_finished
    self._channel.basic_ack(delivery_tag=delivery_tag)
AttributeError: 'NoneType' object has no attribute 'basic_ack'
```
but I suspect that's just a consequence of the first error.

--- a/aiida/work/processes.py
+++ b/aiida/work/processes.py
@@ -16,6 +16,7 @@ import itertools
 import plumpy
 import uuid
 import traceback
+from pika.exceptions import ConnectionClosed
 
 from plumpy import ProcessState
 from aiida.common import exceptions
@@ -160,7 +161,10 @@ class Process(plumpy.Process):
         result = super(Process, self).kill(msg)
 
         for child in self.calc.called:
-            self.runner.rmq.kill_process(child.pk, 'Killed by parent<{}>'.format(self.calc.pk))
+            try:
+                self.runner.rmq.kill_process(child.pk, 'Killed by parent<{}>'.format(self.calc.pk))
+            except ConnectionClosed:
+                self.logger.info('no connection available to kill child<{}>'.format(child.pk))
 
         return result
 

--- a/aiida/work/rmq.py
+++ b/aiida/work/rmq.py
@@ -24,9 +24,29 @@ DeliveryFailed = plumpy.DeliveryFailed
 # know how to avoid warnings. For more info see
 # https://github.com/aiidateam/aiida_core/issues/1142
 _RMQ_URL = 'amqp://127.0.0.1'
+_RMQ_TASK_PREFETCH_COUNT = 20
+_RMQ_HEARTBEAT_TIMEOUT = 600
 _LAUNCH_QUEUE = 'process.queue'
 _MESSAGE_EXCHANGE = 'messages'
 _TASK_EXCHANGE = 'tasks'
+
+
+def get_rmq_url(heartbeat_timeout=None):
+    """
+    Get the URL to connect to RabbitMQ
+
+    :param heartbeat_timeout: the interval in seconds for the heartbeat timeout
+    :returns: the connection URL string
+    """
+    url = _RMQ_URL
+
+    if heartbeat_timeout is None:
+        heartbeat_timeout = _RMQ_HEARTBEAT_TIMEOUT
+
+    if heartbeat_timeout is not None:
+        url += '?heartbeat={}'.format(heartbeat_timeout)
+
+    return url
 
 
 def get_rmq_prefix():
@@ -55,8 +75,9 @@ def get_rmq_config(prefix=None):
         prefix = get_rmq_prefix()
 
     rmq_config = {
-        'url': _RMQ_URL,
+        'url': get_rmq_url(),
         'prefix': prefix,
+        'task_prefetch_count': _RMQ_TASK_PREFETCH_COUNT
     }
 
     return rmq_config
@@ -94,9 +115,8 @@ def get_task_exchange_name(prefix):
 
 def encode_response(response):
     """
-    Used by kiwipy to encode a message for sending.  Because we can have nodes
-    we have to convert these to PIDs before sending (we can't just send the live
-    instance)
+    Used by kiwipy to encode a message for sending.  Because we can have nodes, we have
+    to convert these to PIDs before sending (we can't just send the live instance)
 
     :param response: The message to encode
     :return: The encoded message
@@ -131,6 +151,7 @@ def store_and_serialize_inputs(inputs):
     """
     _store_inputs(inputs)
     return serialize_data(inputs)
+
 
 def _store_inputs(inputs):
     """
@@ -208,7 +229,8 @@ class ProcessControlPanel(object):
             task_queue=task_queue,
             encoder=encode_response,
             decoder=decode_response,
-            testing_mode=testing_mode
+            testing_mode=testing_mode,
+            task_prefetch_count=_RMQ_TASK_PREFETCH_COUNT
         )
 
     def __enter__(self):
@@ -313,7 +335,7 @@ def new_blocking_control_panel():
 def create_rmq_connector(loop=None):
     if loop is None:
         loop = plumpy.events.new_event_loop()
-    return plumpy.rmq.RmqConnector(amqp_url=_RMQ_URL, loop=loop)
+    return plumpy.rmq.RmqConnector(amqp_url=get_rmq_url(), loop=loop)
 
 
 def create_communicator(loop=None, prefix=None, testing_mode=False):
@@ -328,5 +350,6 @@ def create_communicator(loop=None, prefix=None, testing_mode=False):
         connector,
         exchange_name=message_exchange,
         task_queue=task_queue,
-        testing_mode=testing_mode
+        testing_mode=testing_mode,
+        task_prefetch_count=_RMQ_TASK_PREFETCH_COUNT
     )

--- a/aiida/work/runners.py
+++ b/aiida/work/runners.py
@@ -238,14 +238,15 @@ class Runner(object):
         finally:
             runner.close()
 
-    def _setup_rmq(self, url, prefix=None, testing_mode=False):
+    def _setup_rmq(self, url, prefix=None, task_prefetch_count=None, testing_mode=False):
         self._rmq_connector = plumpy.rmq.RmqConnector(amqp_url=url, loop=self._loop)
 
         self._rmq_communicator = plumpy.rmq.RmqCommunicator(
             self._rmq_connector,
             exchange_name=rmq.get_message_exchange_name(prefix),
             task_queue=rmq.get_launch_queue_name(prefix),
-            testing_mode=testing_mode
+            testing_mode=testing_mode,
+            task_prefetch_count=task_prefetch_count
         )
 
         self._rmq = rmq.ProcessControlPanel(
@@ -280,8 +281,8 @@ class DaemonRunner(Runner):
         kwargs['rmq_submit'] = True
         super(DaemonRunner, self).__init__(*args, **kwargs)
 
-    def _setup_rmq(self, url, prefix=None, testing_mode=False):
-        super(DaemonRunner, self)._setup_rmq(url, prefix, testing_mode)
+    def _setup_rmq(self, url, prefix=None, task_prefetch_count=None, testing_mode=False):
+        super(DaemonRunner, self)._setup_rmq(url, prefix, task_prefetch_count, testing_mode)
 
         # Create a context for loading new processes
         load_context = plumpy.LoadContext(runner=self)

--- a/requirements.txt
+++ b/requirements.txt
@@ -143,4 +143,4 @@ Werkzeug==0.14.1
 wrapt==1.10.11
 yapf==0.19.0
 -e git://github.com/muhrin/plumpy.git@4a4b3a49520e3450d49e29b74cc2eb18782c5dd4#egg=plumpy
--e git://github.com/muhrin/kiwipy.git@3ab98fe840d731224571da29df8f155bf6f0d3b8#egg=kiwipy
+-e git://github.com/muhrin/kiwipy.git@561d0a8b97b2a8923e9d85b4cd1c39f0dbe9d182#egg=kiwipy

